### PR TITLE
Hw01

### PR DESCRIPTION
**Pull request by LidaChk**
### HW 01: Introduction, GIT, Collections framework 
# Custom List Implementation
## 2.1 Basic Implementation (Required)
- [x] Create a `CustomList` class that implements the `java.util.List` interface
- [x] Implement a dynamic array (not a linked list) that automatically resizes when capacity is exceeded
- [x] Implement all required methods from the `List` interface
- [x] Focus on core functionality:
  - [x] `add()`
  - [x] `get()`
  - [x] `set()`
  - [x] `remove()`
  - [x] `size()`
  - [x] `isEmpty()`
  - [x] `clear()`
## 2.2 Unit Testing (Required)
- [x] Write comprehensive unit tests for your `CustomList` using JUnit library
- [x] Test basic operations (`add`, `get`, `set`, `remove`)
- [x] Test edge cases:
  - [x] Empty list
  - [x] Boundary indices
  - [x] Null values
- [x] Test list growth and capacity management
- [x] Ensure all tests pass before proceeding
- [x] (*) Write the test in such a way that it verifies the behavior of both `ArrayList` and `CustomList` — they should exhibit identical behavior and produce the same results
## 2.3 Generic Implementation* (Bonus - 1 star)
- [x] Enhance your `CustomList` to support generics
- [x] Allow type-safe operations: `CustomList<Integer>`, `CustomList<String>`, etc.
- [x] Ensure type safety at compile time
- [x] Update your unit tests to work with generic types
## 2.4 Performance Testing** (Bonus - 2 stars)
- [x] Create performance benchmarks comparing your `CustomList` with Java's built-in collections:
### Collections to compare:
- [x] `java.util.ArrayList`
- [x] `java.util.LinkedList`
- [x] Your `CustomList`
#### Bulk Addition Test:
- [x] Add 1,000,000 elements and measure:
  - [x] Execution time
  - [x] Memory usage
#### Add/Remove Test:
- [x] Add 10,000 elements
- [x] Remove the first element 10,000 times and measure:
  - [x] Execution time
  - [x] Memory usage
### Metrics to collect:
- [x] Time taken (milliseconds)
- [x] Memory consumption (MB)
- [x] Create a simple report comparing the results

## 2.5 Complete Implementation*** (Bonus - 3 stars)
- [ ] Implement all List interface methods including:
  - [ ] All collection methods:
    - [ ] `contains()`
    - [ ] `containsAll()`
    - [ ] `addAll()`
    - [ ] `removeAll()`
    - [ ] `retainAll()`
    - [ ] ...
  - [ ] Iterator implementation:
    - [ ] `iterator()`
    - [ ] `listIterator()`
  - [ ] Sublist functionality:
    - [ ] `subList()`
  - [ ] Proper `equals()` and `hashCode()` methods

### CODE REVIEW

File: src/test/java/org/example/week01/ListImplementationTest.java [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
+    @DisplayName("Default constructor should create empty list")
+    Stream<DynamicTest> defaultConstructorShouldCreateEmptyList() {
+      return createTestsForEachList("Default constructor should create empty list", null,
+          (List<String> list) -> {
+            assertEquals(0, list.size());
+            assertTrue(list.isEmpty());
+          });
+    }
+
+    @TestFactory
```
Comment (gneginskiy) [inline src/test/java/org/example/week01/ListImplementationTest.java:63]: quite unsual!
pretty complex abstraction for a simple task 😃 
@ParametrizedTests with methodsource would be easier to understand 

good for now 👍  in future solutions please keep it stupidly simple

**Review by gneginskiy — APPROVED** [review]

File: src/test/java/org/example/week01/ListImplementationTest.java (hunk 1 above)
Comment (LidaChk) [reply src/test/java/org/example/week01/ListImplementationTest.java:63]: Yes, I probably overcomplicated it. I'll try parametric tests in the future. Thank you.
